### DISCUSSION TITLE
Options to reduce Milvus RAM usage?

### DISCUSSION THREAD

**Discussion by P-Ehlert** [milvus-io/milvus#29774] [Q&A and General discussion]
About two years ago I generated a database with Milvus v0.8 standalone containing almost 60 million vectors with dimension 300 and index type ~IVF_FLAT~ IVF_SQ8. This worked fine, but I am now trying to update and regenerate a database using Milvus v2.2 standalone and using varchar(255) as primary key. However, after inserting 47 million vectors the milvus-standalone container crashed. 

The crash seems to be caused by a lack of RAM. The server has 31GB RAM available and just starting Milvus results in about 23GB RAM in use by the milvus-standalone container. When trying to load the collection I see memory increasing to 29GB and then the container crashes. Setting a memory limit of 25GB in the docker-compose.yml file results in an error when trying to load the collection (`Waiting load thread is timeout, loading process may not be finished` and when doing a query `java.lang.Exception: collection:dec2023 or partition:[] not loaded into memory when query`), but at least it prevents the container from crashing.

I'm now looking into options to reduce the RAM consumption. Looking at this page https://milvus.io/tools/sizing/ it's my understanding that with current settings I would actually need ~134.2 GB RAM~ around 50GB which is not an option unfortunately. 

My questions are:
- How come the memory usage of Milvus 2.2. is so much higher than Milvus 0.8?
- What would work better to reduce RAM usage; switching to IVF_SQ8 or IVF_PQ?
- I was reading about on-disk indexing [here](https://milvus.io/docs/v2.2.x/disk_index.md) in the hopes that could help but if I understand that page correctly this is already enabled by default right?
- What else could I do (besides reducing the number of vectors or vector dimension) to reduce RAM usage?

**Comment by yhmo** [comment]
60M entities, 300dim, the raw size of vectors is 60M * 300 * 4bytes = 70GB.
If your primary key is full-length of 255 characters, the raw size of primary keys is 60M * 255bytes = 15GB

So, if you create IVF_FLAT index, the required memory size is 85GB.
The sizing tool estimates the memory size with a safety ratio(2.0) since sometimes the memory peak could be very high.

For IVF_SQ8, the index size is 70GB * 30% = 21GB, the total required memory size is 21+15=33GB, it also exceeds your RAM capacity.
If your primary key is a short string, then it is possible to load the collection with IVF_SQ8 index on your machine.

**Comment by yhmo** [comment]
Milvus v0.8 is written by C/C++, it is a single process, the memory usage can be easily controlled.
Milvus v2.2 is written by Golang + C/C++, there are memory copies in passing data between the two languages. And Milvus v2.2 has several dependencies: Etcd and MinIO, all the dependencies occupy resources.

**Reply by xiaofan-luan** [reply]
The estimation of memory is usually much larger than it actually needed.

For you data size, there is no way you can fit all the data into memory without quantization.

I don't think you can load 60M 300dim data with IVF_FLAT even if you use milvus 0.8 (Unless you enabled mmap features), see the computation of yhmo.

There are certain option if you do want to put all the data with 32G memory:
1. using IVFSQ8 or IVFPQ if you don't care about recall
2. using diskann if your instance has high performance disk and you want high recall
3. using mmap feature in 2.3 - this will slow down you performance dramatically. I guess on milvus 0.8 you are using similar features where your data is swapped between disk and memory so you can fit in large data into your memory

**Reply by P-Ehlert** [reply]
Thank you both for your answers. They helped me understand things a bit better. I do see now that my previous message was incorrect unfortunately. I remembered incorrectly and see now I was already using IVF_SQ8 and nlist 16384 with Milvus 0.8 and the same also with Milvus 2.2. Sorry about this confusion.

I have Milvus v0.8 still up and running and can see RAM consumption there is around 18GB RAM for a database with about 55 million entries. So I'm starting to wonder why I'm running out of memory now with Milvus v2.2 with 48 million items.
Primary key length on average is most likely not more than 60 characters, so I guess memory consumption should be around 
48M * 300 * 4 bytes * 30% = 18GB plus the primary keys 48M * 60 = 3 GB extra = 21 GB
Is there any way to diagnose why memory consumption is actually higher than this?

I also see CPU usage of my Milvus v2.2. spiking for about 15 seconds every minute. Could it be that the system is trying to recover from the previous crash? A lot of logs are being generated and I tried to understand them. What I see is failures trying to load segments (which makes sense because of the lack of memory), but I also see a lot of logs similar to

    {"log":"[2024/01/09 12:32:27.194 +00:00] [DEBUG] [indexcoord/index_coord.go:768] [\"IndexCoord GetIndexInfos\"] [collectionID=445635954798405784] [indexName=] [segIDs=\"[445635954826564726]\"]\n","stream":"stdout","time":"2024-01-09T12:32:27.194875796Z"}
and eventually

    {"log":"[2024/01/09 12:32:27.208 +00:00] [INFO] [datacoord/handler.go:119] [GetQueryVChanPositions] [collectionID=445635954798405784] [channel=by-dev-rootcoord-dml_10_445635954798405784v0] [numOfSegments=333]\n","stream":"stdout","time":"2024-01-09T12:32:27.209080473Z"}
    {"log":"[2024/01/09 12:32:27.209 +00:00] [INFO] [datacoord/handler.go:288] [\"channel seek position set from channel checkpoint meta\"] [channel=by-dev-rootcoord-dml_10_445635954798405784v0] [posTs=446904012062064641] [posTime=2024/01/09 12:31:52.810 +00:00]\n","stream":"stdout","time":"2024-01-09T12:32:27.209224584Z"}
    {"log":"[2024/01/09 12:32:27.209 +00:00] [INFO] [datacoord/services.go:824] [\"datacoord append channelInfo in GetRecoveryInfo\"] [collectionID=445635954798405784] [partitionIDs=\"[]\"] [channelInfo=\"collectionID:445635954798405784 channelName:\\\"by-dev-rootcoord-dml_10_445635954798405784v0\\\" seek_position:\u003cchannel_name:\\\"by-dev-rootcoord-dml_10_445635954798405784v0\\\" msgID:\\\"\\\\210\\\\227\\\\214^\\\\370\\\\2413\\\\006\\\" msgGroup:\\\"by-dev-dataNode-76-by-dev-rootcoord-dml_10_445635954798405784v0\\\" timestamp:446904012062064641 \u003e flushedSegmentIds:445635954850072782 flushedSegmentIds:445635954823571384 flushedSegmentIds:445635954846976187 flushedSegmentIds:445635954828857557 flushedSegmentIds:445635954804138296 flushedSegmentIds:445635954831990419 flushedSegmentIds:445635954852685492 flushedSegmentIds:445635954854277705 flushedSegmentIds:445635954849570119 flushedSegmentIds:445635954847497682 flushedSegmentIds:445635954843228081 flushedSegmentIds:445635954849571000 flushedSegmentIds:445635954854551540 flushedSegmentIds:445635954849058714 flushedSegmentIds:445635954849058682 flushedSegmentIds:445635954840693630 flushedSegmentIds:445635954854286826 flushedSegmentIds:445635954852393830 flushedSegmentIds:445635954851135653 flushedSegmentIds:445635954843227534 flushedSegmentIds:445635954854277109 flushedSegmentIds:445635954852401370 flushedSegmentIds:445635954836658590 flushedSegmentIds:445635954842712890 flushedSegmentIds:445635954853483207 flushedSegmentIds:445635954848539040 flushedSegmentIds:445635954810493875 flushedSegmentIds:445635954817972451 flushedSegmentIds:445635954829972237 flushedSegmentIds:445635954836930795 flushedSegmentIds:445635954850596238 flushedSegmentIds:445635954846480385 flushedSegmentIds:445635954848037604 flushedSegmentIds:445635954849569447 flushedSegmentIds:445635954814231619 flushedSegmentIds:445635954826565600 flushedSegmentIds:445635954816273046 flushedSegmentIds:445635954811398879 flushedSegmentIds:445635954849063965 flushedSegmentIds:445635954808909571 flushedSegmentIds:445635954845224523 flushedSegmentIds:445635954841926483 flushedSegmentIds:445635954807300325 flushedSegmentIds:445635954813719607 flushedSegmentIds:445635954844725711 flushedSegmentIds:445635954849570067 flushedSegmentIds:445635954844726542 flushedSegmentIds:445635954836930726 flushedSegmentIds:445635954835455334 flushedSegmentIds:445635954841919633 flushedSegmentIds:445635954847497581 flushedSegmentIds:445635954844722091 flushedSegmentIds:445635954842720372 flushedSegmentIds:445635954854876247 flushedSegmentIds:445635954836930172 flushedSegmentIds:445635954848539223 flushedSegmentIds:445635954845222441 flushedSegmentIds:445635954848038589 flushedSegmentIds:445635954846510436 flushedSegmentIds:445635954833232744 flushedSegmentIds:445635954828837685 flushedSegmentIds:445635954830724999 flushedSegmentIds:445635954854873009 flushedSegmentIds:445635954849060502 flushedSegmentIds:445635954842710783 flushedSegmentIds:445635954834442042 flushedSegmentIds:445635954827711437 flushedSegmentIds:445635954820359603 flushedSegmentIds:445635954850072966 flushedSegmentIds:445635954836655226 flushedSegmentIds:445635954850599852 flushedSegmentIds:445635954806814111 flushedSegmentIds:445635954853488760 flushedSegmentIds:445635954817976257 flushedSegmentIds:445635954853185252 flushedSegmentIds:445635954825058058 flushedSegmentIds:445635954851140260 flushedSegmentIds:445635954846477033 flushedSegmentIds:445635954817983411 flushedSegmentIds:445635954808913506 flushedSegmentIds:445635954835455346 flushedSegmentIds:445635954848037576 flushedSegmentIds:445635954850600799 flushedSegmentIds:445635954842713148 flushedSegmentIds:445635954851136777 flushedSegmentIds:445635954827709004 flushedSegmentIds:445635954851135646 flushedSegmentIds:445635954851425353 flushedSegmentIds:445635954825045280 flushedSegmentIds:445635954823577495 flushedSegmentIds:445635954825040957 flushedSegmentIds:445635954847503187 flushedSegmentIds:445635954835460716 flushedSegmentIds:445635954841165889 flushedSegmentIds:445635954825064879 flushedSegmentIds:445635954806098556 flushedSegmentIds:445635954834223874 flushedSegmentIds:445635954850601808 flushedSegmentIds:445635954854549498 flushedSegmentIds:445635954837135418 flushedSegmentIds:445635954854876656 flushedSegmentIds:445635954841162245 flushedSegmentIds:445635954811635848 flushedSegmentIds:445635954848541055 flushedSegmentIds:445635954825045640 flushedSegmentIds:445635954818860115 flushedSegmentIds:445635954833233384 flushedSegmentIds:445635954835450459 flushedSegmentIds:445635954820358855 flushedSegmentIds:445635954846976448 flushedSegmentIds:445635954800291806 flushedSegmentIds:445635954827694651 flushedSegmentIds:445635954802421522 flushedSegmentIds:445635954843970454 flushedSegmentIds:445635954849570992 flushedSegmentIds:445635954853187306 flushedSegmentIds:445635954851141391 flushedSegmentIds:445635954840686690 flushedSegmentIds:445635954852680679 flushedSegmentIds:445635954851424759 flushedSegmentIds:445635954847499578 flushedSegmentIds:445635954813734103 flushedSegmentIds:445635954842712905 flushedSegmentIds:445635954829956250 flushedSegmentIds:445635954807772023 flushedSegmentIds:445635954845735534 flushedSegmentIds:445635954832440653 flushedSegmentIds:445635954831986658 flushedSegmentIds:445635954846477013 flushedSegmentIds:445635954843970653 flushedSegmentIds:445635954841162231 flushedSegmentIds:445635954854555571 flushedSegmentIds:445635954853486558 flushedSegmentIds:445635954834439255 flushedSegmentIds:445635954854557202 flushedSegmentIds:445635954842720347 flushedSegmentIds:445635954846471091 flushedSegmentIds:445635954834437402 flushedSegmentIds:445635954834223800 flushedSegmentIds:445635954853488392 flushedSegmentIds:445635954848037663 flushedSegmentIds:445635954802401107 flushedSegmentIds:445635954842710684 flushedSegmentIds:445635954844725313 flushedSegmentIds:445635954843429275 flushedSegmentIds:445635954846975089 flushedSegmentIds:445635954841926076 flushedSegmentIds:445635954853485152 flushedSegmentIds:445635954841926031 flushedSegmentIds:445635954833229871 flushedSegmentIds:445635954848541080 flushedSegmentIds:445635954843965357 flushedSegmentIds:445635954806814104 flushedSegmentIds:445635954850596285 flushedSegmentIds:445635954851425222 flushedSegmentIds:445635954834442045 flushedSegmentIds:445635954849061030 flushedSegmentIds:445635954845737843 flushedSegmentIds:445635954826571032 flushedSegmentIds:445635954852397242 flushedSegmentIds:445635954841926091 flushedSegmentIds:445635954808913496 flushedSegmentIds:445635954835450536 flushedSegmentIds:445635954845225012 flushedSegmentIds:445635954847499551 flushedSegmentIds:445635954827708560 flushedSegmentIds:445635954806826535 flushedSegmentIds:445635954806344424 flushedSegmentIds:445635954840695033 flushedSegmentIds:445635954811638862 flushedSegmentIds:445635954841165111 flushedSegmentIds:445635954851430558 flushedSegmentIds:445635954839874921 flushedSegmentIds:445635954852984066 flushedSegmentIds:445635954846972339 flushedSegmentIds:445635954841926505 flushedSegmentIds:445635954811398906 flushedSegmentIds:445635954835460707 flushedSegmentIds:445635954834437429 flushedSegmentIds:445635954841165197 flushedSegmentIds:445635954843965390 flushedSegmentIds:445635954843968784 flushedSegmentIds:445635954847499587 flushedSegmentIds:445635954849569487 flushedSegmentIds:445635954835439015 flushedSegmentIds:445635954848036899 flushedSegmentIds:445635954845741986 flushedSegmentIds:445635954804870717 flushedSegmentIds:445635954820363361 flushedSegmentIds:445635954854876393 flushedSegmentIds:445635954804623683 flushedSegmentIds:445635954836653908 flushedSegmentIds:445635954849566020 flushedSegmentIds:445635954853186132 flushedSegmentIds:445635954848539180 flushedSegmentIds:445635954844726498 flushedSegmentIds:445635954834439262 flushedSegmentIds:445635954831228934 flushedSegmentIds:445635954831989141 flushedSegmentIds:445635954841926120 flushedSegmentIds:445635954809570628 flushedSegmentIds:445635954816248948 flushedSegmentIds:445635954847499622 flushedSegmentIds:445635954854279835 flushedSegmentIds:445635954849060520 flushedSegmentIds:445635954851136738 flushedSegmentIds:445635954855365851 flushedSegmentIds:445635954848539052 flushedSegmentIds:445635954842981314 flushedSegmentIds:445635954817980526 flushedSegmentIds:445635954800287518 flushedSegmentIds:445635954816013978 flushedSegmentIds:445635954800285165 flushedSegmentIds:445635954833233502 flushedSegmentIds:445635954848038634 flushedSegmentIds:445635954826564726 flushedSegmentIds:445635954811180618 flushedSegmentIds:445635954844764619 flushedSegmentIds:445635954848540642 flushedSegmentIds:445635954848540687 flushedSegmentIds:445635954846511043 flushedSegmentIds:445635954850599815 flushedSegmentIds:445635954852681757 flushedSegmentIds:445635954849565980 flushedSegmentIds:445635954820358926 flushedSegmentIds:445635954850068680 flushedSegmentIds:445635954850601858 flushedSegmentIds:445635954831224852 flushedSegmentIds:445635954843964330 flushedSegmentIds:445635954816000508 flushedSegmentIds:445635954852399157 flushedSegmentIds:445635954850074139 flushedSegmentIds:445635954836934618 flushedSegmentIds:445635954851425465 flushedSegmentIds:445635954854283875 flushedSegmentIds:445635954842979502 flushedSegmentIds:445635954833232716 flushedSegmentIds:445635954843228180 flushedSegmentIds:445635954855368476 flushedSegmentIds:445635954846481585 flushedSegmentIds:445635954837830848 flushedSegmentIds:445635954823565871 flushedSegmentIds:445635954848036951 flushedSegmentIds:445635954853771739 flushedSegmentIds:445635954836930212 flushedSegmentIds:445635954825052096 flushedSegmentIds:445635954842978972 flushedSegmentIds:445635954851425237 flushedSegmentIds:445635954850068694 flushedSegmentIds:445635954853483418 flushedSegmentIds:445635954823564477 flushedSegmentIds:445635954849060998 flushedSegmentIds:445635954813969755 flushedSegmentIds:445635954845735511 flushedSegmentIds:445635954846481696 flushedSegmentIds:445635954825041561 flushedSegmentIds:445635954828856602 flushedSegmentIds:445635954855368380 flushedSegmentIds:445635954846480334 flushedSegmentIds:445635954844722034 flushedSegmentIds:445635954804390430 flushedSegmentIds:445635954840692985 flushedSegmentIds:445635954853490558 flushedSegmentIds:445635954841919593 flushedSegmentIds:445635954835439089 flushedSegmentIds:445635954829970854 flushedSegmentIds:445635954831990423 flushedSegmentIds:445635954850074144 flushedSegmentIds:445635954844725724 flushedSegmentIds:445635954831986632 flushedSegmentIds:445635954851422879 flushedSegmentIds:445635954817976985 flushedSegmentIds:445635954809358301 flushedSegmentIds:445635954801923473 flushedSegmentIds:445635954833214832 flushedSegmentIds:445635954829970126 flushedSegmentIds:445635954848037698 flushedSegmentIds:445635954850072839 flushedSegmentIds:445635954839869535 flushedSegmentIds:445635954828855518 flushedSegmentIds:445635954855365292 flushedSegmentIds:445635954852395974 flushedSegmentIds:445635954853481318 flushedSegmentIds:445635954851137357 flushedSegmentIds:445635954837135473 flushedSegmentIds:445635954846508886 flushedSegmentIds:445635954836934630 flushedSegmentIds:445635954839867746 flushedSegmentIds:445635954825056821 flushedSegmentIds:445635954844725386 flushedSegmentIds:445635954843968823 flushedSegmentIds:445635954845737800 flushedSegmentIds:445635954811180608 flushedSegmentIds:445635954807521928 flushedSegmentIds:445635954842983103 flushedSegmentIds:445635954839861573 flushedSegmentIds:445635954846504039 flushedSegmentIds:445635954806584441 flushedSegmentIds:445635954813719511 flushedSegmentIds:445635954831188246 flushedSegmentIds:445635954831989154 flushedSegmentIds:445635954844765146 flushedSegmentIds:445635954833229859 flushedSegmentIds:445635954854277058 flushedSegmentIds:445635954854283916 flushedSegmentIds:445635954811635845 flushedSegmentIds:445635954845742409 flushedSegmentIds:445635954851140261 flushedSegmentIds:445635954813734204 flushedSegmentIds:445635954845224098 flushedSegmentIds:445635954844766151 flushedSegmentIds:445635954850600786 flushedSegmentIds:445635954846471125 flushedSegmentIds:445635954808675788 flushedSegmentIds:445635954843964304 flushedSegmentIds:445635954804374621 flushedSegmentIds:445635954841165234 flushedSegmentIds:445635954842713182 flushedSegmentIds:445635954841165972 flushedSegmentIds:445635954841165125 flushedSegmentIds:445635954800291972 flushedSegmentIds:445635954844763364 flushedSegmentIds:445635954833214949 flushedSegmentIds:445635954801431948 flushedSegmentIds:445635954850072756 flushedSegmentIds:445635954851142466 flushedSegmentIds:445635954851137299 flushedSegmentIds:445635954845741958 flushedSegmentIds:445635954849063954 flushedSegmentIds:445635954809358293 flushedSegmentIds:445635954831224412 flushedSegmentIds:445635954845742449 flushedSegmentIds:445635954847502970 \"]\n","stream":"stdout","time":"2024-01-09T12:32:27.209512979Z"}

Does this mean the system is still busy flushing inserted entries into the database? I remember that with Milvus v0.8 we turned off WAL to prevent issues generating the database, but I don't see such an option for Milvus 2.2. 
Anyway, any help into this problem is very much appreciated.

I appended a part of the logs here: [milvus22logs.txt.gz](https://github.com/milvus-io/milvus/files/13874335/milvus22logs.txt.gz)

**Reply by yhmo** [reply]
```
{"log":"[2024/01/09 13:20:18.784 +00:00] [INFO] [querynode/segment_loader.go:890] [\"predict memory and disk usage while loading (in MiB)\"] [collectionID=445635954798405784] [concurrency=1] [memLoadingUsage=23335] [memUsageAfterLoad=23099] [diskUsageAfterLoad=0] [currentUsedMem=23021] [currentAvailableFreeMemory=2578] [currentTotalMemory=25600]\n","stream":"stdout","time":"2024-01-09T13:20:18.784993973Z"}
{"log":"[2024/01/09 13:20:18.784 +00:00] [ERROR] [querynode/segment_loader.go:154] [\"load failed, OOM if loaded\"] [collectionID=445635954798405784] [segmentType=Sealed] [\"loadSegmentRequest msgID\"=2202] [error=\"InsufficientMemoryToLoad, load segment failed, OOM if load, collectionID = 445635954798405784, MaxSegmentMemUsage = 236 MB, concurrency = 1, usedMemAfterLoad = 23099 MB, currentAvailableMemCount = 2578 MB, currentTotalMem = 25600 MB, thresholdFactor = 0.900000\"] [stack=\"github.com/milvus-io/milvus/internal/querynode.(*segmentLoader).LoadSegment\\n\\t/go/src/github.com/milvus-io/milvus/internal/querynode/segment_loader.go:154\\ngithub.com/milvus-io/milvus/internal/querynode.(*loadSegmentsTask).Execute\\n\\t/go/src/github.com/milvus-io/milvus/internal/querynode/load_segment_task.go:81\\ngithub.com/milvus-io/milvus/internal/querynode.(*taskScheduler).processTask\\n\\t/go/src/github.com/milvus-io/milvus/internal/querynode/task_scheduler.go:108\\ngithub.com/milvus-io/milvus/internal/querynode.(*taskScheduler).taskLoop\\n\\t/go/src/github.com/milvus-io/milvus/internal/querynode/task_scheduler.go:130\"]\n","stream":"stdout","time":"2024-01-09T13:20:18.785172219Z"}
{"log":"[2024/01/09 13:20:18.784 +00:00] [WARN] [querynode/load_segment_task.go:125] [\"failed to load segment\"] [collectionID=445635954798405784] [replicaID=446786465552662529] [error=\"InsufficientMemoryToLoad, load segment failed, OOM if load, collectionID = 445635954798405784, MaxSegmentMemUsage = 236 MB, concurrency = 1, usedMemAfterLoad = 23099 MB, currentAvailableMemCount = 2578 MB, currentTotalMem = 25600 MB, thresholdFactor = 0.900000\"]\n","stream":"stdout","time":"2024-01-09T13:20:18.785217632Z"}
{"log":"[2024/01/09 13:20:18.785 +00:00] [WARN] [querynode/task_scheduler.go:110] [\"InsufficientMemoryToLoad, load segment failed, OOM if load, collectionID = 445635954798405784, MaxSegmentMemUsage = 236 MB, concurrency = 1, usedMemAfterLoad = 23099 MB, currentAvailableMemCount = 2578 MB, currentTotalMem = 25600 MB, thresholdFactor = 0.900000\"]\n","stream":"stdout","time":"2024-01-09T13:20:18.785231302Z"}
```


currentTotalMem = 25600 MB  ----------------------------------------------------- Although your machine has 32GB RAM. the milvus container is limited to use no more than 24GB RAM. The milvus server predicts memory usage during loading the segments one by one. For each segment, it estimates the memory peak by segment_size * 3. And there is a configurable threshold for memory usage, the default value is 90%. That is to say, if the estimated memory usage exceeds `currentTotalMem * 90%`, it will fail to load. 

Remove `deploy.resources.limits.memory` in the docker_compose.yaml. If you are using MAC OS, double check your docker configuration, enlarge the RAM limit size:
![mac docker](https://github.com/milvus-io/milvus/assets/2282099/b8c7a9c4-153b-4e36-bf85-04b383f7936c)

**Reply by P-Ehlert** [reply]
Thank you for your answer. I added the limit on purpose because without it the Milvus container crashed. I could set it a bit higher, but since there is another process running on the same machine (to insert the vectors) I cannot go higher than 28GB.
Do you think the OOM is also causing the large number of other logs that Milvus is generating. I got the impression those are unrelated?

**Comment by xiaofan-luan** [comment]
I'd recommend to use latest 2.2.16 or 2.3.4 because we did many optimization for memory usage.
v2.2 did have some waste on memory and it's not stable enough as well

**Reply by P-Ehlert** [reply]
Thanks, I'll try upgrading to 2.3.4 then. Hopefully your optimizations will keep the memory low enough.
I think I'll also experiment with IVF_PQ to see if that also helps keep RAM usage down a bit.
